radix-engine/src/blueprints/resource/resource_manager_common.rs

```diff
@@ -221,35 +259,98 @@ fn build_access_rules(
             ObjectModuleId::SELF,
             FUNGIBLE_VAULT_LOCK_FUNGIBLE_AMOUNT_IDENT,
         ),
-        AllowAll,
+        AccessRuleEntry::AccessRule(AccessRule::Protected(AccessRuleNode::ProofRule(
+            ProofRule::Require(SoftResourceOrNonFungible::StaticNonFungible(
+                NonFungibleGlobalId::package_actor(RESOURCE_MANAGER_PACKAGE),
+            )),
+        ))),
         DenyAll,
     );
     vault_access_rules.set_method_access_rule_and_mutability(
         MethodKey::new(
             ObjectModuleId::SELF,
             NON_FUNGIBLE_VAULT_LOCK_NON_FUNGIBLES_IDENT,
         ),
-        AllowAll,
+        AccessRuleEntry::AccessRule(AccessRule::Protected(AccessRuleNode::ProofRule(
+            ProofRule::Require(SoftResourceOrNonFungible::StaticNonFungible(
+                NonFungibleGlobalId::package_actor(RESOURCE_MANAGER_PACKAGE),
+            )),
+        ))),
         DenyAll,
     );
     vault_access_rules.set_method_access_rule_and_mutability(
         MethodKey::new(
             ObjectModuleId::SELF,
             FUNGIBLE_VAULT_UNLOCK_FUNGIBLE_AMOUNT_IDENT,
         ),
-        AllowAll,
+        AccessRuleEntry::AccessRule(AccessRule::Protected(AccessRuleNode::ProofRule(
+            ProofRule::Require(SoftResourceOrNonFungible::StaticNonFungible(
+                NonFungibleGlobalId::package_actor(RESOURCE_MANAGER_PACKAGE),
+            )),
+        ))),
         DenyAll,
     );
     vault_access_rules.set_method_access_rule_and_mutability(
         MethodKey::new(
             ObjectModuleId::SELF,
             NON_FUNGIBLE_VAULT_UNLOCK_NON_FUNGIBLES_IDENT,
         ),
-        AllowAll,
+        AccessRuleEntry::AccessRule(AccessRule::Protected(AccessRuleNode::ProofRule(
+            ProofRule::Require(SoftResourceOrNonFungible::StaticNonFungible(
+                NonFungibleGlobalId::package_actor(RESOURCE_MANAGER_PACKAGE),
+            )),
+        ))),
         DenyAll,
     );
 
-    (resman_access_rules, vault_access_rules)
+    // Not that if a local reference to a bucket is passed to another actor, the recipient will be able
+    // to take resource from the bucket. This is not what Scrypto lib supports/encourages, but can be done
+    // theoretically.
```

Comment:
I think we will want to prevent local references from being passed through call frames.

Reply:
This is probably part of the larger discussion around locks/refs/call frames.

Reply:
Exactly!